lib/Infer/AliveDriver.cpp

```diff
@@ -0,0 +1,411 @@
+      auto i = std::make_unique<IR::Input>(t, std::move(x));
+      auto ptr = i.get();
+      F.addInput(std::move(i));
+      identifiers[x] = ptr;
+      return ptr;
+    }
+  }
+
+  IR::Value *ToValue(IR::Type& t, const char *x) {
+    return ToValue(t, std::string(x));
+  }
+
+  IR::Value *ToValue(IR::Type& t, IR::Value *v) {
+    return v;
+  }
+};
+
+
+bool StartsWith(const std::string &pre, const std::string &str) {
+  return std::equal(pre.begin(), pre.end(), str.begin());
+}
+
+void GetReservedConsts(souper::Inst *I, std::set<std::string>& Result) {
+  if (StartsWith("reservedconst_", I->Name)) {
+    Result.insert(I->Name);
+  }
+  for (auto &&Op : I->Ops) {
+    GetReservedConsts(Op, Result);
+  }
+}
+
+struct ReturnLHSRAII {
+  tools::Transform& t;
+  IR::Function &LHS;
+  ~ReturnLHSRAII() {
+    LHS = std::move(t.src);
+  }
+};
+}
+
+util::Errors SynthesizeConstantUsingSolver(tools::Transform &t,
+                                std::string ConstName, int64_t &result) {
+
+  IR::Value::reset_gbl_id();
+  IR::State src_state(t.src), tgt_state(t.tgt);
+  util::sym_exec(src_state);
+  util::sym_exec(tgt_state);
+
+  util::Errors errs;
+
+  auto &ap = src_state.returnVal();
+  auto &bp = tgt_state.returnVal();
+  auto &a = ap.first;
+  auto &b = bp.first;
+
+  auto qvars = src_state.getQuantVars();
+  qvars.insert(ap.second.begin(), ap.second.end());
+
+  auto err = [&](const smt::Result &r, bool print_var, const char *msg) {
+    tools::error(errs, src_state, tgt_state, r, print_var, nullptr, a, b, msg,
+          false);
+  };
+
+  std::set<smt::expr> vars;
+  std::set<smt::expr> consts;
+
+  for (auto &[var, val] : src_state.getValues()) {
+    auto &name = var->getName();
+    if (StartsWith("%var", name)) {
+      auto app = val.first.value.isApp();
+      assert(app);
+      vars.insert(Z3_get_app_arg(smt::ctx(), app, 1));
+    }
+  }
+  for (auto &[var, val] : tgt_state.getValues()) {
+    auto &name = var->getName();
+    if (StartsWith("%reserved", name)) {
+      auto app = val.first.value.isApp();
+      assert(app);
+      consts.insert(Z3_get_app_arg(smt::ctx(), app, 1));
+    }
+  }
+
+  auto SimpleConstExistsCheck = smt::expr::mkForAll(vars, a.value == b.value);
+
+  smt::Solver::check({{preprocess(t, qvars, ap.second,
+                std::move(SimpleConstExistsCheck)),
+  [&] (const smt::Result &r) {
+    if (r.isUnsat()) {
+      err(r, true, "Value mismatch");
+    } else if (r.isSat()) {
+      result = r.getModel().getInt(*consts.begin());
+      return;
+    } else {
+      err(r, true, "Unknown/Invalid Result, investigate.");
+    }
+  }}});
+
+  return errs;
+}
+
+souper::AliveDriver::AliveDriver(Inst *LHS_, Inst *PreCondition_)
+    : LHS(LHS_), PreCondition(PreCondition_) {
+  InstNumbers = 101;
+
+  if (!TranslateRoot(LHS, PreCondition, LHSF, LExprCache)) {
+    llvm::errs() << "Failed to translate LHS.\n";
+    assert(false);
+  } else {
+    LHSFExists = true;
+  }
+}
+
+//TODO: Maybe return an APInt
```

Comment:
yes, it's definitely not ok to limit constants to 64 bits

Reply:
smt::Model::getInt returns an int64, so the limiting is not being done here.  It might still make sense to use APInt here for future proofing.

Reply:
It seems I can get a Z3 expr from the model and use its string representation to construct an APInt.
Should this go into Alive instead of Souper?  
I want to avoid using Z3 directly as much as possible here.

Reply:
I'm afraid consulting the model directly will not help, I looked into it a bit and the 64-bit limit in alive2 is pervasive and there's no point fixing this until the problems are fixed